### PR TITLE
Lane Following and Line Detection 

### PR DESCRIPTION
**Pull request by raadkhan**
Using Sliding Window + PolyFit method

https://chatbotslife.com/self-driving-cars-advanced-computer-vision-with-opencv-finding-lane-lines-488a411b2c3d

### CODE REVIEW

File: src/better_lane_follow/src/LineDetect.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ * Subscribes to:
+ * Publishes to:
+ */
+
+#include <LineDetect.h>
+#include <opencv2/imgproc/imgproc.hpp>
+#include <opencv2/highgui/highgui.hpp>
+#include <opencv2/objdetect/objdetect.hpp>
+
+LineDetect::LineDetect() : initalLineDetectThreshold(200),
```
Comment (ValRat) [inline src/better_lane_follow/src/LineDetect.cpp]: These are simple integers, so I don't see a need for an initialisation list - these are generally used if you're constructing more complicated members of a class. Also, `initalLineDetectThreshold(200)` is misspelt.

File: src/better_lane_follow/src/LineDetect.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            }
+        }
+    }
+
+    return histogram;
+}
+
+std::vector<Polynomial2D> getLines(cv::Mat& binaryImage) {
+
+    std::vector<int> baseHistogram = getHistogram(binaryImage);
```
Comment (ValRat) [inline src/better_lane_follow/src/LineDetect.cpp]: Where is `getHistogram` defined? Needs to be implemented

File: src/better_lane_follow/src/LineDetect.cpp (hunk 1 above)
Comment (garethellis0) [reply src/better_lane_follow/src/LineDetect.cpp]: How would you suggest initialising them? I'm not super particular about it, but I feel like this is a decent format that forces all the variable initialisation to be in one place. Certainly I've seen integers initialised in production codebases like this.

File: src/better_lane_follow/src/LineDetect.cpp (hunk 2 above)
Comment (garethellis0) [reply src/better_lane_follow/src/LineDetect.cpp]: `LineDetect::getHistogram` is defined above, but this function should be `LineDetect::getLines`, not just `getLines`

Comment (raadkhan) [reply src/better_lane_follow/src/LineDetect.cpp]: resolved

File: src/better_lane_follow/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,69 @@
+cmake_minimum_required(VERSION 2.8.3)
+project(better_lane_follow)
```
Comment (garethellis0) [inline src/better_lane_follow/CMakeLists.txt]: We should probably change the name to just `lane_following`, seeing as it looks like we're just going to toss our first attempt

File: src/better_lane_follow/include/LineDetect.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Created by: Raad Khan
+ * Created On: April 23, 2017
+ * Description: Takes in an image feed and generates lane lines.
+ * Usage:
+ * Subscribes to:
+ * Publishes to:
```
Comment (garethellis0) [inline src/better_lane_follow/include/LineDetect.h]: I'm OK if we remove these two lines. It's likely that subscription and publish topics will change, and these will not, just causing confusion down the line.

File: src/better_lane_follow/package.xml [hunk 5]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0"?>
+<package>
+  <name>better_lane_follow</name>
```
Comment (garethellis0) [inline src/better_lane_follow/package.xml]: if we change the package name as per my comment above, will probably have to change it here as well

File: src/better_lane_follow/src/LineDetect.cpp (hunk 1 above)
Comment (garethellis0) [reply src/better_lane_follow/src/LineDetect.cpp]: @ValRat ?

File: src/better_lane_follow/test/my_node_rostest.cpp [hunk 6]
```diff
@@ -0,0 +1,68 @@
+/*
+ * Created By: Valerian Ratu
+ * Created On: January 29, 2017
+ * Description: Integration testing for MyNode
+ */
```
Comment (garethellis0) [inline src/better_lane_follow/test/my_node_rostest.cpp]: Not sure if we're going to have time for rostests along with the unit tests, so let's just remove this file for now. I think it's just copied over from the `IGVC-2017` repo, so we should be able to easily re-add it if needed.

File: src/better_lane_follow/test/sample_package_test.test [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<!-- A .test file which launches the node and its corresponding rostest -->
+
+<launch>
+    <node name="my_node" pkg="sample_package" type="my_node" />
+    <test test-name="my_node_rostest" pkg="sample_package" type="my_node_rostest" />
+</launch>
```
Comment (garethellis0) [inline src/better_lane_follow/test/sample_package_test.test]: If we're getting rid of the rostests for now, we can also get rid of this file.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Some general notes:
- I think we probably want to remove the `lane_follow` package, and rename `better_lane_follow` to `lane_following` (or something like that). The joy of git is that we can always get files we delete back (even deletions are just changes that can be reverted later down the line), so don't worry too much about losing stuff.

File: src/better_lane_follow/src/LineDetect.cpp [hunk 8]
```diff
@@ -47,34 +46,88 @@ std::vector<Polynomial2D> getLines(cv::Mat& binaryImage) {
...
-            window.center = peak.second + (window.center - window.width/2);
+            intVec windowHistogram = LineDetect::getHistogram(windowSlice);
+            std::pair<int, int> peak = LineDetect::getHistogramPeak(windowHistogram);
 
+            window.center = peak.second + (window.center - window.width/2);
         }
     }
 
-    std::vector<Polynomial2D> polyLines;
+    std::vector<Polynomial2D> polyLines = LineDetect::contructPolyLine(linePoints, degree);
```
Comment (raadkhan) [inline src/better_lane_follow/src/LineDetect.cpp]: first parameter should actually be ```points```

File: src/better_lane_follow/src/LineDetect.cpp (hunk 1 above)
Comment (raadkhan) [reply src/better_lane_follow/src/LineDetect.cpp]: resolved

File: src/better_lane_follow/include/LineDetect.h (hunk 4 above)
Comment (raadkhan) [reply src/better_lane_follow/include/LineDetect.h]: resolved

File: src/better_lane_follow/test/my_node_rostest.cpp (hunk 6 above)
Comment (raadkhan) [reply src/better_lane_follow/test/my_node_rostest.cpp]: resolved

File: src/better_lane_follow/test/sample_package_test.test (hunk 7 above)
Comment (raadkhan) [reply src/better_lane_follow/test/sample_package_test.test]: resolved

File: src/better_lane_follow/CMakeLists.txt (hunk 3 above)
Comment (raadkhan) [reply src/better_lane_follow/CMakeLists.txt]: done

File: src/better_lane_follow/package.xml (hunk 5 above)
Comment (raadkhan) [reply src/better_lane_follow/package.xml]: done

**Pull request closed without merging**